commons-math/src/userguide/java/org/apache/commons/math4/userguide/genetics/PolygonChromosome.java

```diff
@@ -0,0 +1,135 @@
+
+    /** The reference image for fitness testing. */
+    private static BufferedImage refImage;
+    /** The image buffer used to draw the current chromosome during fitness testing. */
+    private static BufferedImage testImage;
+
+    public static void setRefImage(BufferedImage ref) {
+        refImage = ref;
+    }
+
+    public static void setTestImage(BufferedImage image) {
+        testImage = image;
+    }
+
+    public PolygonChromosome(List<Polygon> representation) {
+        super(representation);
+    }
+
+    @Override
+    protected void checkValidity(List<Polygon> chromosomeRepresentation) throws InvalidRepresentationException {
+        // do nothing
+    }
+
+    @Override
+    public AbstractListChromosome<Polygon> newFixedLengthChromosome(List<Polygon> chromosomeRepresentation) {
+        return new PolygonChromosome(chromosomeRepresentation);
+    }
+
+    /**
+     * Return the internal representation, which is needed for our custom mutation policy.
+     *
+     * @return the list of polygons
+     */
+    public List<Polygon> getPolygonRepresentation() {
+        return getRepresentation();
+    }
+
+    /**
+     * Calculate the fitness function for this chromosome.
+     * <p>
+     * For this purpose, we first draw the polygons on the test buffer, and
+     * then compare the resulting image pixel by pixel with the reference image.
+     */
+    public double fitness() {
+
+        Graphics2D g2 = testImage.createGraphics();
+        
+        int width = testImage.getWidth();
+        int height = testImage.getHeight();
+
+        draw(g2, width, height);
+        g2.dispose();
+
+        int[] refPixels = refImage.getData().getPixels(0, 0, refImage.getWidth(), refImage.getHeight(), (int[]) null);
+        int[] testPixels = testImage.getData().getPixels(0, 0, testImage.getWidth(), testImage.getHeight(), (int[]) null);
+
+        int diff = 0;
+        int p = width * height * 4 - 1; // 4 channels: rgba
+        int idx = 0;
+
+        do {
+            if (idx++ % 4 != 0) { // ignore the alpha channel for fitness
+                int dp = testPixels[p] - refPixels[p];
+                if (dp < 0) {
+                    diff -= dp;
+                } else {
+                    diff += dp;
+                }
+            }
+        } while(--p > 0);
+
+        return (1.0 - diff / (width * height * 3.0 * 256));
+    }
+
+    public void draw(Graphics2D g, int width, int height) {
+        g.setBackground(Color.WHITE);
+        g.clearRect(0, 0, width, height);
+
+        g.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
+        g.setComposite(AlphaComposite.getInstance(AlphaComposite.SRC_OVER));
+
+        List<Polygon> polygons = getPolygonRepresentation();
+        for (Polygon p : polygons) {
+            p.draw(g, width, height);
+        }
+    }
+
+    @Override
+    public String toString() {
+        return String.format("(f=%s)", getFitness());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format double: &#39;getFitness&#39; expression. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 124. Consider formatting this data with &#39;%f&#39; instead.